src/metabase/notification/send.clj

```diff
@@ -188,12 +181,89 @@
+  NotificationQueueProtocol
+  (put-notification! [_ notification]
+    (let [id (or (:id notification) (str (random-uuid)))]
+      (.lock queue-lock)
+      (try
+        (when-not (.contains ids-list id)
+          (.add ids-list id))
+        (.put id->notification id notification)
+        ;; Signal that a notification is available
+        (.signal not-empty-cond)
+        (finally
+          (.unlock queue-lock)))))
+
+  (take-notification! [_]
+    (.lock queue-lock)
+    (try
+      ;; Wait until there's at least one notification
+      (while (.isEmpty ids-list)
+        (.await not-empty-cond))
+      (let [id           (.removeFirst ids-list)
+            notification (.remove id->notification id)]
+        notification)
+      (finally
+        (.unlock queue-lock)))))
+
+(defn- create-notification-queue
+  "A thread-safe, notification queue with the following properties:
+  - Notifications are identified by unique IDs
+  - Adding a notification with an ID already in the queue replaces the existing one
+  - Taking from an empty queue blocks until a notification is available
+  - Multiple threads can safely add and take from the queue concurrently"
+  []
+  (let [queue-lock     (java.util.concurrent.locks.ReentrantLock.)
+        not-empty-cond (.newCondition queue-lock)]
+    (->NotificationQueue (java.util.LinkedList.)
+                         (java.util.concurrent.ConcurrentHashMap.)
+                         queue-lock
+                         not-empty-cond)))
+
+(defn- create-notification-dispatcher
+  "Create a thread pool for sending notifications.
+  There can only be one notification with the same id in the queue.
+  - if a notification of the same id is already in the queue, then replace it
+  - if a notification doesn't have id, put it into queue regardless (used to send unsaved notifications)"
+  [pool-size queue]
+  (let [executor (Executors/newFixedThreadPool
+                  pool-size
+                  (.build
+                   (doto (BasicThreadFactory$Builder.)
+                     (.namingPattern "send-notification-thread-pool-%d"))))]
+    (.addShutdownHook
+     (Runtime/getRuntime)
+     (Thread. ^Runnable (fn []
+                          (.shutdown ^ExecutorService executor))))
+    (dotimes [_ pool-size]
+      (.submit ^ExecutorService executor
+               ^Callable (fn []
+                           (while true
+                             (try
+                               (let [notification (take-notification! queue)]
+                                 (send-notification-sync! notification))
+                               (catch Exception e
+                                 (log/error e "Error in notification worker")))))))
+    (fn [notification]
+      (put-notification! queue notification))))
+
+(defonce ^:private dispatcher
+  (delay (create-notification-dispatcher (notification-thread-pool-size) (create-notification-queue))))
```

Comment:
What do you think about hoisting the queue up into a var, so we can inspect or manipulate it in the REPL? A hazard comment warning people from using it directly should be enough... a linter feels like too much work.

Reply:
did this.